### PR TITLE
Fix: extract method for SWD parity calculation

### PR DESCRIPTION
**Pull request by ALTracer**
## Detailed description

This PR is aimed at two problems:
* Problem 1: swdptap of firmware references both popcount and parity builtins to do the same task (of calculating SWD parity), and `__popcountsi2(): 40 bytes` is more expensive. Solution 1: create a shim/dispatcher in maths_utils, name it `swd_odd_parity` and call *that* from swdptap (the gpio-bitbanging one). This leads to cleaner code at that callsite, and less flash space occupied
* Problem 2: *_swd.c `hosted` adapter driver backends also want to calculate parity, and also do that via both popcount and parity builtins. MSVC does not provide these builtins, only the `__popcount` for x86-64 optional instruction. Solution 2: incorporate this _MSC-specific builtin as well into the shim, and tweak all said `hosted` parity calculations to call the shim.

This PR is triggered from the discussion of #1688 two weeks ago.
I decided to name the function `swd_odd_parity()` for where it is going to be used. If future JTAG-PDI code also would want to use it (and attribute alias does not cut it), the reviewer/implementer of that can suggest a better name and I'll mass-rename the touched callsites (with a force-push).
Notice that I drop `maths_utils.h` from `general.h` and include it where needed by this PR. The point on making the shim `static inline` in the header is a valid discussion topic.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`)
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues
May help fix building BMDA under MSVC with `HOSTED_BMP_ONLY=0` (caveat libusb/hidapi).

### CODE REVIEW

File: src/platforms/hosted/ftdi_swd.c
Comment (dragonmux) [inline src/platforms/hosted/ftdi_swd.c:241]: Given the check in `ftdi_swd_seq_in_parity` makes it impossible for clock_cycles to exceed 32, this can be made a uint32_t and the `UINT64_C(1)` switched to `1U`. If this causes a build error, cast `clock_cycles` to inhibit that.

File: src/platforms/common/swdptap.c
Comment (dragonmux) [inline src/platforms/common/swdptap.c:144]: Rather than using `!=`, what does the codegen look like for `^`? If memory serves, the original `+=` pattern (which, granted, should have been `^=`) produces slightly fewer instructions than the comparison, so we'd be interested to see what different operations for calculating if the parity matches or doesn't does to this

File: src/maths_utils.c
Comment (dragonmux) [inline src/maths_utils.c:90]: This needs to be `result ^= (value >> bit) & 1U;`, otherwise it expresses a very slow bitwise copy of `value` to `result` as `^=` is bitwise, not logical.

File: src/include/maths_utils.h
Comment (dragonmux) [inline src/include/maths_utils.h:40]: The `const` here can be dropped (clang-tidy should be linting about this) from here (but not the definition in maths_utils.c).

Perhaps lets use `calc_odd_parity` or `calculate_odd_parity` as a more generic term for what this is doing as it's non-specific to SWD and that'll keep things open in the nomenclature for the JTAG-PDI support which will definitely benefit from this?

File: src/maths_utils.c
Comment (dragonmux) [inline src/maths_utils.c:88]: With the change below, this can become a `uint8_t` or a `bool` - your choice - and the result type changed to match.

**Review by dragonmux — CHANGES_REQUESTED** [review]
It's definitely looking good - nice job cleaning things up and rationalising them with this change!

We've spotted a couple of functional issues in review, noted below, and have a couple of name suggestions to pick from for what to name this new function, but we're happy once those are addressed to get this merged.

It's definitely a shame we don't have a nice way of building a sim suite or test suite for the maths functions so we can have in-repo proofs of their correctness, but that's an improvement to strive for in a future PR.

File: src/maths_utils.c
Comment (ALTracer) [reply src/maths_utils.c:88]: Most hosted code uses `uint8_t` for packets represented as arrays. I'm more inclined to swap the `bool` in common/swdptap.c to uint8_t, or even `uint32_t`, which is faster (no bitmasking in codegen).
The original libgcc2 `__builtin_parity()` and `__builtin_popcount()` take `unsigned int` and return `int`, which is why I picked `uint32_t` to be transparent enough in the shim case. MSVC docs say `unsigned int __popcnt(unsigned int value);`, supporting this. In the "generic impl" case (on exotic toolchains?) I think we still should aim to return the same "int". Of course, if BMD only requires a 1 or a 0 from it, then it can be demoted to `uint8_t` everywhere, unless the hosted code already does this.

File: src/maths_utils.c
Comment (ALTracer) [reply src/maths_utils.c]: Hate bitwise loops. Fixed.

File: src/maths_utils.c
Comment (dragonmux) [reply src/maths_utils.c:88]: The function can only produce the values '0' and '1', so hence suggesting using `uint8_t` or `bool` - given that point about uint8_t for packets and the fact this get used to initialise values in those packets, it supports going to `uint8_t` and doing all the type truncation here - this will eliminate a pile of `-Wconversion`/`-Warith-conversion` warnings and we don't think will negatively impact codegen.

File: src/include/maths_utils.h
Comment (ALTracer) [reply src/include/maths_utils.h:40]: `calculate_odd_parity` is good, but too verbose -- all math functions "calculate". To help IntelliSense and likes, I suggest `parity_odd` to put the context first. I wonder if `parity_even` exists elsewhere.
The `const`, to my understanding, should be visible from headers so that the compiler can optimize other translation units to not duplicate arguments into r0 (or whatever the AAPCS first argument register is). Or are we only using it for guarantees on pointers?

Comment (dragonmux) [reply src/include/maths_utils.h:40]: `parity_odd` reads.. strangely.. to us, due to us referencing here calculating the parity value for an odd parity scheme - it's technically correct, but it's got the same odd lilt as Picard saying "Tea, Earl Grey, Hot, Black" rather than "a hot cup of black Earl Grey tea".

The `const` on a value in a C function declaration (prototype) does nothing, and is purely noise - this is what the clang-tidy lint is about. It only has meaning in the function definition.

In regards pointers, `const char *` is saying the thing the pointer points to is const, but `const char *const` doesn't do anything extra in the context of a decl. That's why we omit on the parameter side of the * in the decl.

File: src/platforms/common/swdptap.c
Comment (ALTracer) [reply src/platforms/common/swdptap.c:144]: Comparing two bools here can be done with `!=` or `^` -- both compile to `eor.w r0, r1, r2;` right before function epilogue `add sp, #16; pop {r4, r5, r6, pc}` or `ldmia.w sp!, {clobbers, pc}`.
This block is cherry-picked from PR1688 for types and const.

Comment (dragonmux) [reply src/platforms/common/swdptap.c:144]: Oh, perfect, in that case.. let's stick with the `!=` check as it's more obvious what it does - a better expression of intent. Glad the compiler's seeing sense and doing the good optimisation there.

File: src/include/maths_utils.h
Comment (ALTracer) [reply src/include/maths_utils.h]: Renamed `swd_odd_parity` to `calculate_odd_parity` globally.
Removed `const` from uint32_t in this header.

File: src/maths_utils.c
Comment (ALTracer) [reply src/maths_utils.c]: Adjusted to `uint8_t`.

**Comment by ALTracer** [comment]
Requesting next round of review.
Fixup squashing order: keep first 5 commits, incorporate last 4 commits into them: 7-8 into 1st, 6/9 into 4-5. On the other hand, I can replay this manually to make it clean.

File: src/platforms/hosted/jlink_swd.c
Comment (dragonmux) [inline src/platforms/hosted/jlink_swd.c:217]: This comparison should probably be converted to `!=` from `^` for clarity now

**Review by dragonmux — COMMENTED** [review]
This all looks good to us now - the one review note here is an optional change. Please replay the history to make it clean and we'll get this merged.

**Comment by ALTracer** [comment]
Merged fixups, propagated the rename, build-tested intermediate commits with a custom pre-commit extended hook, and rebased to latest main.
Think that's ready for final live testing for SWD parity regressions and any remaining review that the individual commits adhere to expectations. Please try it on a Tigard or something else FTDI.
There's one last `__builtin_ctz` remaining in ftdi.c:`ftdi_swd_init()`, and I have some ideas on how to go around it, but maybe not in this PR -- and I'd like MSVC build logs w.r.t. that.

**Comment by dragonmux** [comment]
Most excellent! Regarding MSVC, once the new Meson support lands it should become considerably easier to generate full BMDA builds under MSVC and get logs from them.

**Review by dragonmux — DISMISSED** [review]
LGTM, approved pending testing with our Tigard. We'll merge (assuming it all works properly, but we believe it should) once that's been tested.

File: src/platforms/hosted/ftdi_swd.c
Comment (dragonmux) [reply src/platforms/hosted/ftdi_swd.c]: Turns out this was wrong, we'd put that `UINT64_C()` in because the bit mask uses a 64-bit intermediary - the switch to `uint32_t` is fine, but `1U << 32U` is generated when `clock_cycles` is 32 which has a 64-bit result and invokes undefined behaviour if attempted in 32-bit space. Found while testing on our Tigard.

**Review by dragonmux — APPROVED** [review]

**Comment by dragonmux** [comment]
We'll get this merged now it's been tested and the one bug found and fixed, good work! 🙂

**Pull request merged**
